sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java

```diff
@@ -957,15 +957,27 @@ public void sessionCreated(IoSession session) throws Exception {
 
         @Override
         public void sessionClosed(IoSession session) throws Exception {
-            TcpipClientChannel channel = (TcpipClientChannel) session.removeAttribute(TcpipClientChannel.class);
+            TcpipClientChannel channel = (TcpipClientChannel) session.removeAttribute(TcpipClientChannel.class);                      
             Throwable cause = (Throwable) session.removeAttribute(TcpipForwardingExceptionMarker.class);
             if (channel != null) {
                 if (debugEnabled) {
                     log.debug("sessionClosed({}) closing channel={} after {} messages - cause={}",
-                            session, channel, messagesCounter, (cause == null) ? null : cause.getClass().getSimpleName());
+                        session, channel, messagesCounter, (cause == null) ? null : cause.getClass().getSimpleName());
+                }
+              
+                if (cause != null) {
+                    channel.close(true);
+                } else {
+                    OpenFuture openFuture = channel.getOpenFuture();
+                    if (!openFuture.isDone()) {
```

Comment:
1.  The remark is misplaced - should be over `channel.close(true)`
2.  I am not comfortable with the idea of delaying the channel closure until the open is finished. This is `sessionClosed` so why wait for the channel open to finish ?

Reply:
1. I agree. I removed the comment and made a new commit.
2. I do not agree. If you don't delay closing the channel until the channel is completely opened, you risk loosing all the data bytes sent by the client to the local port forwarding listening socket/tunnel. This is what happened in our test case.

Reply:
I am inclined to agree with you on (2) -  reviewing and testing...